### ISSUE TITLE
Inplace operation error in gradient computation

### ISSUE THREAD

**Issue by cgsavard** [hqucms/weaver#13]
![Screen Shot 2022-09-28 at 5 53 57 PM](https://user-images.githubusercontent.com/15660721/192908503-d40e8b15-ea93-4f2a-bd68-5c9d38c38b2b.png)

I have come across this error when trying to train. After a bit of google searching, it seems that something is being updated during the gradient computation before it should be. I was hoping you could help me locate the error and let me know what I need to fix as I am not too familiar with pytorch. I have made no modifications to the utils/nn/tools.py script.

**Comment by cgsavard** [comment]
I have solved the issue by changing all the inplace operations [here](https://github.com/hqucms/weaver/blob/master/utils/nn/model/ParticleNet.py#L183-L184) and [here](https://github.com/hqucms/weaver/blob/master/utils/nn/model/ParticleNet.py#L265-L272) to non-inplace. Essentially the change is var1 \*= var2 was changed to var1 = var1\*var2. Should this be changed in the code permanently to avoid this error in the future?

**Comment by hqucms** [comment]
Hi @cgsavard -- can you share the pytorch version? I don't seem to be able to reproduce this error in e.g., `1.12.1`.

**Comment by cgsavard** [comment]
Yes, this occurred after I installed pytorch in this way "conda install pytorch torchvision torchaudio cudatoolkit=11.6 -c pytorch -c conda-forge" because I was working on a different GPU (K40) with a version of CUDA >11.6 (11.7). The pytorch version was the stable 1.12.1 and so I think it was actually the newer CUDA version that raised the issue.

**Comment by hqucms** [comment]
I tested CUDA 11.6 + PyTorch 1.12.1 and still cannot reproduce this error. Did you change anything else when the problem got solved?
